src/metabase/driver/sql/query_processor.clj

```diff
@@ -44,9 +44,11 @@
   nil)
 
 (defn make-nestable-sql
-  "For embedding native sql queries, wraps [sql] in parens and removes any semicolons"
+  "For embedding native sql queries, wraps [sql] in parens and removes any comments and semicolons."
   [sql]
-  (str "(" (str/replace sql #";[\s;]*$" "") ")"))
+  (str "(" (-> sql
+               (str/replace #"--.*(\n|$)" "")
```

Comment:
Will this remove something like `select 'not -- a comment' from x` ? 
It would feel safer to check something like `^\s*--.*(\n|$)` but I'm not sure that fully captures the problem. 
I think `select a from x; -- comment prevents semicolon removal` is the gist right?
In that case, I think we're out of regex into sql parsing territory. I wonder if a better error message that checks for `--` in a `catch` if a query fails with a syntax error would be sufficient to point people in the right direction until we tackle parsing.

Reply:
> Will this remove something like `select 'not -- a comment' from x` ?

I've tweaked the `make-nestable-sql` so it's bit more robust, but still it does just a regex replacement. Please check the code.

> In that case, I think we're out of regex into sql parsing territory.

Probably. Another approach how to tackle this could be to write a tokenizer that handles string literals, identifiers, semicolons, comments - which could be simpler than parsing. But making that work across all supported sql dialects would require addition of driver specific rulesets eg. as driver methods.

Mock with hardcoded "best effort" ruleset:

```clojure
(defn- tokenize-sql
    "Build vector of tokens from sql string for purposes of removal traling comments and semicolons."
    [sql]
    (loop [acc []
           str sql]
      (if (empty? str)
        acc
        (let [[type token]
              (condp re-find str
                #"^'(?:''|.|\n)*?'"        :>> #(vector :quoted %)
                #"^\"(?:.|\n)*?\""         :>> #(vector :double-quoted %)
                #"^`(?:.|\n)*?`"           :>> #(vector :backticked %)
                #"^--.*"                   :>> #(vector :comment %)
                #"^;+"                     :>> #(vector :semi %)
                #"^\s+"                    :>> #(vector :space %)
                #"^(?:\n|.)+?(?=`|\"|'|--|;|$)" :>> #(vector :rest %)
                (throw (ex-info "Failed to tokenize sql string"
                                {:sql sql
                                 :remaining str
                                 :tokenized acc})))]
          (assert (not-empty token) "Failed to tokenize sql string")
          (recur (conj acc [type token]) (subs str (count token)))))))
          
(defn remove-trailing-comments-and-semis
    "Remove trailing semicolons and comments from sql string."
    [sql]
    (let [tokens (tokenize-sql sql)
          trailing-removed (->> (rseq tokens)
                                (drop-while (fn [[type]] (#{:comment :semi :space} type)))
                                reverse)]
      (->> trailing-removed
           (map (fn [[_ token]] token))
           (str/join ""))))
```


> I wonder if a better error message that checks for `--` in a `catch` if a query fails with a syntax error would be sufficient to point people in the right direction until we tackle parsing.

As the problem is subquery specific, we would have to save information like "this is a risky query" somewhere, when subquery containing `;\s*--` is encountered eg. during source query fetching or parameter substitution -- prior to `make-nestable-sql` call. And then use it when execution of the root query fails. It would not be enough to just check the root query after the execution as `;\s*--` patterns would be already removed.